### ISSUE TITLE
[Bug]: Babel removes parenthesis at place where are required.

### ISSUE THREAD

**Issue by gabrielgortabns** [babel/babel#16111]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```js
import something from 'something';

export function myfunction(bool) {
    return /** @type {SomeType} */ (
        bool ? something.a() : something.b()
    )
}
```

### Current and expected behavior

```js
const something = require('something');

exports.myfunction = myfunction;

function myfunction(bool) {
    return; /** @type {SomeType} */ 
        bool ? something.a() : something.b()
}
```

### Environment

  System:
    OS: Windows 11 10.0.22621
  Binaries:
    Node: 18.6.0 - C:\Program Files\nodejs\node.EXE
    npm: 8.15.0 - C:\Program Files\nodejs\npm.CMD
  npmPackages:
    @babel/cli: ^7.23.0 => 7.23.0 
    @babel/core: ^7.23.2 => 7.23.2 
    @babel/plugin-transform-modules-commonjs: ^7.23.3 => 7.23.3 
    @babel/preset-env: ^7.23.3 => 7.23.3 
    @babel/traverse: ^7.23.3 => 7.23.3 
    eslint: ^8.53.0 => 8.53.0 
    jest: ^29.7.0 => 29.7.0 

### Additional context

The issue happens when there is `/** @type {SomeType} */` in return statement, or in assigment. Causing either in invalid return, or even if code itself works, in output files, TypeScript throws error because of invalid types (as without parenthesis it doesn't apply type cast on whole expression).

### reproduction in comment below:
https://github.com/babel/babel/issues/16111#issuecomment-1825331645

**Comment by nicolo-ribaudo** [comment]
I'm releasing a patch right now that might fix this issue

**Comment by liuxingbaoyu** [comment]
I'm worried this PR might not be helpful.
Also can you provide a way to reproduce it? I can't reproduce it in repl.

[repl](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=JYWwDg9gTgLgBAZwiApjAFsAdgczgMymTgHIlUNscSBuAKDpQA9JYCBXLAYxmAizggAnvk48-WABQAjCBAA2ASjgBvOnA1woadlAEB6AFSG4AARhCwKVQGVkKACqWUAXziH9cSes2_ZCuAB-RHtKXAA6AENJZQAuEIpMCOkYnw1FOhcgA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=true&sourceType=module&lineWrap=false&presets=&prettier=false&targets=&version=7.23.3&externalPlugins=%40babel%2Fplugin-transform-modules-commonjs%407.23.3&assumptions=%7B%7D
)

**Comment by GabrielGorta** [comment]
Checking tests in PR #16104 and it really seems it will fix the issue.

**Comment by liuxingbaoyu** [comment]
That test isn't actually what's being fixed, it's a test of https://github.com/babel/babel/pull/16104, which I'm worried has a regression.

**Comment by kungfooman** [comment]
Isn't this because `createParenthesizedExpressions` is `false` by default (for whatever reason, AST vs. CST story)?

I wanted to test this in REPL, but I can't find a way to set `{createParenthesizedExpressions: true}`. Is there any way to set `parse` options in REPL?

**Comment by liuxingbaoyu** [comment]
> Isn't this because `createParenthesizedExpressions` is `false` by default (for whatever reason, AST vs. CST story)?
> 
> I wanted to test this in REPL, but I can't find a way to set `{createParenthesizedExpressions: true}`. Is there any way to set `parse` options in REPL?

No, even this is unexpected when `creatingParenthesizedExpressions: false`. This is different in AST.

> I wanted to test this in REPL, but I can't find a way to set {createParenthesizedExpressions: true}. Is there any way to set parse options in REPL?

Unfortunately there is currently no easy way.

**Comment by liuxingbaoyu** [comment]
You can view them via `astexplorer`. 
https://astexplorer.net/#/gist/c8fa290417825f5efb7f858da5be9f14/28b66f5d63d27c90887e83d09ab237b9d9aa34e1

**Comment by kungfooman** [comment]
@liuxingbaoyu Thank you, I mean integrated into the PR build, because some plugin's or options could manipulate the AST (e.g. `createParenthesizedExpressions`)

**Comment by bjornharrtell** [comment]
This appears to be cause of https://github.com/openlayers/openlayers/issues/15358. A reproduction can be found at https://github.com/bjornharrtell/ol-issue-15358.

**Comment by ahocevar** [comment]
> This appears to be cause of https://github.com/openlayers/openlayers/issues/15358. A reproduction can be found at https://github.com/bjornharrtell/ol-issue-15358.

Looks like this is a regression in "@babel/parser". Downgrading to `@babel/parser@7.19.0` makes the problem go away.

**Comment by GabrielGorta** [comment]
@bjornharrtell

Amazing job! I've just donwgraded the `@babel/parser` and now it works correctly at least in runtime (e.g. return statement). It's exactly same issue. Thank you @ahocevar too, for finding solution.

**Comment by kungfooman** [comment]
@ahocevar I tried to reproduce the issue and it's kinda confusing:

```sh
mkdir babel_paren
cd babel_paren
```

Create file **test.mjs**:

```js
import {parse} from '@babel/parser';
import generator from '@babel/generator';
const generate = generator.default;
const code = `
function getResolutionForValueFunction(power) {
  power = power || 2;
  const maxResolution = this.getConstrainedResolution(this.maxResolution_);
  const minResolution = this.minResolution_;
  const max = Math.log(maxResolution / minResolution) / Math.log(power);
  return (
    /**
     * @param {number} value Value.
     * @return {number} Resolution.
     */
    function (value) {
      const resolution = maxResolution / Math.pow(power, value * max);
      return resolution;
    }
  );
}
`;
const ast = parse(code);
const out = generate(ast,
  {
    compact: true,
    //createParenthesizedExpressions: true,
  }
);
console.log(out);
```

First run:

```sh
rm -rf node_modules package.json package-lock.json && npm i @babel/parser@7.23.3 @babel/generator@7.23.3 && node test.mjs
```

```
added 12 packages, and audited 13 packages in 737ms

found 0 vulnerabilities
{
  code: 'function getResolutionForValueFunction(power){power=power||2;const maxResolution=this.getConstrainedResolution(this.maxResolution_);const minResolution=this.minResolution_;const max=Math.log(maxResolution/minResolution)/Math.log(power);return(/**\n' +
    '     * @param {number} value Value.\n' +
    '     * @return {number} Resolution.\n' +
    '     */function(value){const resolution=maxResolution/Math.pow(power,value*max);return resolution;});}',
  decodedMap: undefined,
  __mergedMap: [Getter],
  map: [Getter/Setter],
  rawMappings: [Getter/Setter]
}
```

Nice parentheses, right?

But now run:

```sh
npm i @babel/generator && node test.mjs
```

Result

```js
up to date, audited 13 packages in 439ms

found 0 vulnerabilities
{
  code: 'function getResolutionForValueFunction(power){power=power||2;const maxResolution=this.getConstrainedResolution(this.maxResolution_);const minResolution=this.minResolution_;const max=Math.log(maxResolution/minResolution)/Math.log(power);return/**\n' +
    '     * @param {number} value Value.\n' +
    '     * @return {number} Resolution.\n' +
    '     */function(value){const resolution=maxResolution/Math.pow(power,value*max);return resolution;};}',
  decodedMap: undefined,
  __mergedMap: [Getter],
  map: [Getter/Setter],
  rawMappings: [Getter/Setter]
}
``

Parentheses are gone! :eyes: 

This only happens when using `{compact: true}` as generate option.

**Comment by liuxingbaoyu** [comment]
@ahocevar @kungfooman 
Thank you for your detailed report!
I think there are two problems here.

https://github.com/babel/babel/issues/16111#issuecomment-1825514962 
Regression from `@babel/generator@7.23.4`, I can reproduce it, will fix it.


And https://github.com/bjornharrtell/ol-issue-15358 is related to `@babel/parser`.
I get the error, but even if I downgrade `@babel/parser` it doesn't work.🤔

**Comment by liuxingbaoyu** [comment]
I don't understand why downgrading `@babel/parer` would work and I can't replicate this locally.
The bug in https://github.com/bjornharrtell/ol-issue-15358 disappeared after I downgraded `@babel/generator`, and it was caused by the merged PR a few days ago.

**Comment by bjornharrtell** [comment]
Confirmed that downgrading to @babel/generator@7.23.0 resolves https://github.com/bjornharrtell/ol-issue-15358. Note that downgrading to @babel/generator@7.23.3 does not appear to be enough.

**Comment by liuxingbaoyu** [comment]
I downgraded it to `@babel/generator@7.23.3` and it worked, this should be fixed soon.

**Comment by Doyler123** [comment]
FYI for anyone that might have started seeing the below error in their react application over the last week, this bug is the cause and it is fixed in `@babel/generator@7.23.5`

`Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=f`

**Comment by bjornharrtell** [comment]
Confirmed @babel/generator@7.23.5 resolves the issue and a lockfile regen/deps cleanup will bring that for me as transitive dependency.
